packages/jaeger-ui/src/components/TracePage/CriticalPath/index.tsx

```diff
@@ -0,0 +1,161 @@
+  const sanitizedSpanData: Span[] = [];
+
+  spans.forEach(span => {
+    if (span.references.length) {
+      const parentSpan = span.references.filter(ref => ref.refType === 'CHILD_OF')[0].span!;
+      const childEndTime = span.startTime + span.duration;
+      const parentEndTime = parentSpan.startTime + parentSpan.duration;
+      if (span.startTime >= parentSpan.startTime && childEndTime <= parentEndTime) {
+        sanitizedSpanData.push(span);
+      } else if (
+        span.startTime < parentSpan.startTime &&
+        childEndTime <= parentEndTime &&
+        childEndTime > parentSpan.startTime
+      ) {
+        sanitizedSpanData.push({
+          ...span,
+          startTime: parentSpan.startTime,
+          duration: childEndTime - parentSpan.startTime,
+        });
+      } else if (
+        span.startTime >= parentSpan.startTime &&
+        childEndTime > parentEndTime &&
+        span.startTime < parentEndTime
+      ) {
+        sanitizedSpanData.push({ ...span, duration: parentEndTime - span.startTime });
+      }
+    } else {
+      sanitizedSpanData.push(span);
+    }
+  });
+
+  return sanitizedSpanData;
+};
+
+export const computeCriticalPath = (
+  traceData: Trace,
+  spanId: string,
+  criticalPath: criticalPathSection[],
+  spawnTime?: number
+) => {
+  // spawnTime refers to spawn synchronization event of current span
+  const currentSpan: Span = traceData.spans.filter(span => span.spanID === spanId)[0];
+
+  const lastFinishingChildSpanId = findLastFinishingChildSpanId(traceData, currentSpan, spawnTime);
+  let spanCriticalSection: criticalPathSection;
+
+  if (lastFinishingChildSpanId) {
+    const lfcSpan = traceData.spans.filter(span => span.spanID === lastFinishingChildSpanId)[0];
+    spanCriticalSection = {
+      spanId: currentSpan.spanID,
+      section_start: lfcSpan.startTime + lfcSpan.duration,
+      section_end: spawnTime || currentSpan.startTime + currentSpan.duration,
+    };
+    criticalPath.push(spanCriticalSection);
+    // Now lfc turns to the currentspan and we again do the same algorithm
+    computeCriticalPath(traceData, lastFinishingChildSpanId, criticalPath);
+  } else {
+    // If there is no last finishing child then total section upto startTime of span is on critical path
+    spanCriticalSection = {
+      spanId: currentSpan.spanID,
+      section_start: currentSpan.startTime,
+      section_end: spawnTime || currentSpan.startTime + currentSpan.duration,
+    };
+    criticalPath.push(spanCriticalSection);
+    // Now as there are no lfc's it goes back to parent span from startTime(spawn Event)
+    if (currentSpan.references.length) {
+      const parentSpan: string = currentSpan.references.filter(
+        reference => reference.refType === 'CHILD_OF'
+      )[0].spanID;
+      computeCriticalPath(traceData, parentSpan, criticalPath, currentSpan.startTime);
+    }
+  }
+  return criticalPath;
+};
+
+function TraceCriticalPath(traceProp: Tprops) {
```

Comment:
this is a function that provides business logic, not a UI component. The "props" nomenclature is not relevant here.